### DISCUSSION TITLE
Implementing data store pattern with Provider?

### DISCUSSION THREAD

**Discussion by tonnyavery** [rrousselGit/riverpod#2085] [General]
Let's say I have a local storage where I store something, for example values user have entered into a list.

Here is what I have tried initially:

```
final tagsDataStoreProvider = Provider(TagsDataStore.new);

class TagsDataStore {
  final Ref ref;
  final List<Tag> _tags;

  TagsDataStore(
    this.ref,
  ) : _tags = [];

  Future<List<Tag>> get() async => _tags

  add(Tag tag) async {
    if (_tags.contains(tag)) return;
    _tags.add(tag);
  }

  edit(Tag tag) async {
    remove(tag);
    _tags.add(tag);
  }

  remove(Tag tag) async {
    _tags.removeWhere((element) => element.name == tag.name);
  }
}

```

But what is not clear is how do I get notified about the change.
If there is some other provider that watches `tagsDataStoreProvider`, it would want to be notified every time new tag is added to the TagsDataStore. But because it's a wrapper around a regular class, it's unclear how can this class notify the provider that contents have changed and it's time to trigger the chain of events that will lead to UI updates.

Reading through docs here https://docs-v2.riverpod.dev/docs/concepts/why_immutability#using-immutable-state
would `StateNotifierProvider`+`StateNotifier` be a correct solution for such data store?

**Comment by tonnyavery** [comment]
I am reading this and examples are a bit confusing: https://docs-v2.riverpod.dev/docs/providers/notifier_provider
With `AsyncNotifierProvider`, what's the pattern for state modification? `AsyncValue.guard` assumes that full new state is delivered, do I need to store my separate state variable for the values?
What if the patten is keep a local cache for quick data and fire to the (remote) database to save data there?

**Reply by andrezanna** [reply]
Hello @AAverin, did you ever find a solution? I'm stuck in a similar situation and not sure how to handle this.

**Reply by tonnyavery** [reply]
@andrezanna Hey.
Not sure if this is the right approach, but here is what I ended up doing to workaround.
DataStore layer is built using regular classes and a provider.
```
final myDataStoreProvider = Provider((ref) => MyDataStore());

class MyDataStore {
  final List<MyData> _datas = [];

  Future<List<MyData>> get() async {
    return _datas;
  }

  Future<MyData> getMyData(String id) async =>
      _datas.firstWhere((element) => element.id == id);

  add(MyData myData) async {
    _datas.add(myData);
  }

  edit(MyData myData) async {
    await remove(myData);
    await add(myData);
  }

  Future<bool> remove(MyData myData) async {
    _datas.removeWhere((element) => element.id == myData.id);
    return true;
  }
}
```

Repository layer is built using AsyncNotifier and AsyncNotifierProvider, so they will notify downstream.
Repositories can use DataStores to store something. 

There is also `ref.invalidate(provider)` method that I use once in a while to invalidate values.
I have wrapped some of the DataStores with a FamilyProvider that would read some specific values I need and I invalidate these family providers to get fresh data.

again, maybe there is a better way to structure all these providers, I wish documentation would be a bit more clear and have more practical examples

**Reply by andrezanna** [reply]
Thanks for your answer, not still what I'm looking for... I need to do more reasearch.

For a similar approach i used a Map instead of a list so i can crud the data easily. Maybe this approach can help you.
(I don't know too if this can be considered a good approach)
```

final exercisesProvider = ChangeNotifierProvider<ExercisesProvider>((ref) {
  return ExercisesProvider();
});

class ExercisesProvider with ChangeNotifier {
  Map<int, Exercise> exercises = {};

  add(Exercise exercise) {
    try {
      exercises[exercise.id] != null;
    } catch (e) {
      exercises[exercise.id] = exercise;
    }
    notifyListeners();
  }

  addAll(List<Exercise> exercises) {
    this.exercises.addEntries(exercises.map((e) => MapEntry(e.id, e)));
    notifyListeners();
  }

  del(Exercise exercise) {
    exercises.remove(exercise.id);
    notifyListeners();
  }

  clear() {
    exercises.clear();
    notifyListeners();
  }

  getExercise(id) async {
    Exercise? exercise;
    try {
      exercise = exercises[id]!;
    } catch (e) {
      exercise = await exerciseRepo.getExercise(id);
      if (exercise != null) {
        exercises[exercise.id] = exercise;
      }
    }
    return exercise;
  }

  refreshExercise(id) async {
    Exercise? exercise;
    try {
      exercise = await exerciseRepo.getExercise(id);
      if (exercise != null) {
        exercises[exercise.id] = exercise;
      }
    } catch (e) {
      exercise = exercises[id]!;
    }
    return exercise;
  }
}

```
